src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SkeletonReferenceCache.cs

```diff
@@ -59,11 +61,20 @@ private partial class SkeletonReferenceCache
         private static readonly EmitOptions s_metadataOnlyEmitOptions = new(metadataOnly: true);
 
         /// <summary>
-        /// Lock we take before emitting metadata.  Metadata emit is extremely expensive.  So we want to avoid cases
-        /// where N threads come in and try to get the skeleton for a particular project.  This way they will instead
-        /// yield if something else is computing and will then use the single instance computed once one thread succeeds.
+        /// Static conditional mapping from a compilation to the skeleton set produced for it.  This is valuable for a
+        /// couple of reasons. First, a compilation tracker may fork, but produce the same compilation.  As such, we
+        /// want to get the same skeleton set for it.  Second, consider the following scenario:
+        /// <list type="number">
+        /// <item>Project A is referenced by projects B and C (both have a different language than A).</item>
+        /// <item>Producing the compilation for 'B' produces the compilation for 'A' which produces the skeleton that 'B' references.</item>
+        /// <item>B's compilation is released and then GC'ed.</item> 
+        /// <item>Producing the compilation for 'C' needs the skeleton from 'A'</item>
+        /// </list>
+        /// At this point we would not want to re-emit the assembly metadata for A's compilation.  We already did that
+        /// for 'B', and it can be enormously expensive to do so again.  So as long as A's compilation lives, we really
+        /// want to keep it's skeleton cache around.
         /// </summary>
-        private readonly SemaphoreSlim _emitGate = new(initialCount: 1);
+        private static readonly ConditionalWeakTable<Compilation, AsyncLazy<SkeletonReferenceSet?>> s_compilationToSkeletonSet = new();
```

Comment:
no need for the gate.  the CWT is already thread-safe on access, and the use of AsyncLazy means there's at most one computation ever happening to produce the set.